### ISSUE TITLE
New API features and redesign

### ISSUE THREAD

**Issue by Freedos** [lformella/xdcc-grabscher#60]
#26 #52 #59

To download packages over the API it would be nice to post a JSON encoded request. For example:

POST ht*p://your-own-host:5556/api/downloadPacket

``` JSON
{
"key": "XXXXX-XXXXX-XXXX-XXXX-XXXXXXXXXX",
"name": "newestshit.tar",
"server": "irc.abjects.net",
"channel": "#mg",
"bot": "somebot",
"id": "#3"
}
```

returns:

``` JSON
    {"ReturnValue":-1} - api key is invalid or disabled
    {"ReturnValue":0} - there was an error calling the method
    {"ReturnValue":1} - everything is fine
```

Also searching for Packets would be nice since SQLite won't be used anymore:

POST ht*p://your-own-host:5556/api/searchPackets

``` JSON
{
"key": "XXXXX-XXXXX-XXXX-XXXX-XXXXXXXXXX",
"included": "Searchterm",
"excluded": "excludedSerchterm"
}
```

which returns a list with JSON encoded Packets

``` JSON
{
"name": "newestshit.tar",
"server": "irc.abjects.net",
"channel": "#mg",
"bot": "somebot",
"id": "#3",
"size": "100 MB",
"speed": "1.16 M",
"lastmentioned": "2014-05-25T21:29:11"
}
{
...
}
```

Please give your suggestions and thoughts on this

**Comment by lformella** [comment]
Please take a look here: https://github.com/lformella/xdcc-grabscher/blob/xg_3.3/README.md#api
The api methods **downloadPacket**, **searchPackets** and **enable** are already implemented.

**Comment by Freedos** [comment]
More than enough for me. I will test it as soon XG 3.3 is released

**Comment by scottc** [comment]
If you are going to redesign the Web API, I suggest following a modified RESTful pattern.
https://en.wikipedia.org/wiki/Representational_State_Transfer#Applied_to_web_services

**Objects:**
Server
Channel
Bot
Package
QueuedPackets

**Methods for each collection / object:**
List/fetch (GET / or /guid)
Replace/Update (PUT / or /guid)
Create (POST / or /guid)
Delete (DELETE / or /guid)

**Formats:**
JSON (default, if not specified)
XML (for legacy support, of xml only parser apps)
HTML (for human consumtion)

**Responses:**
200 OK
404 not found
500 Server Error (see body for message, or HTTP status message)
etc

```
GET /api/server.json
```

returns list of servers in json format

```
GET /api/server.html
```

returns list of servers in html format

```
GET /api/package/5e5903f3-71ab-41fd-919e-bdadbfd04abf.xml
```

returns a single package, in xml format.

```
GET /api/queuedpackage.json
GET /api/queuedpackage/5e5903f3-71ab-41fd-919e-bdadbfd04abf.html
PUT /api/queuedpackage/5e5903f3-71ab-41fd-919e-bdadbfd04abf
POST /api/queuedpackage/5e5903f3-71ab-41fd-919e-bdadbfd04abf
DELETE /api/queuedpackage/5e5903f3-71ab-41fd-919e-bdadbfd04abf
POST /api/channel
POST /api/channel/5e5903f3-71ab-41fd-919e-bdadbfd04abf
PUT /api/bot
PUT /api/bot/5e5903f3-71ab-41fd-919e-bdadbfd04abf
DELETE /api/package
DELETE /api/package/5e5903f3-71ab-41fd-919e-bdadbfd04abf
```

etc

Additionally the magical xdcc api, that is nested and a bit more smarter by following the existing xdccLink convention as well, which will automatically create all parent objects if they do not exist.

```
GET /api/xdcc/server/channel/bot/package/queuedpackage
POST /api/xdcc/server/channel/bot/package/queuedpackage
PUT /api/xdcc/server/channel/bot/package/queuedpackage
DELETE /api/xdcc/server/channel/bot/package/queuedpackage
```

So in theory, one could create all objects needed and add the package to the queue to be downloaded, in a single HTTP request.

For deleting and creating, it would delete/update the lowest level item or collection specified, in the url request.

Or something similar.

**Comment by lformella** [comment]
I updated the api again to match the rest standard: https://github.com/lformella/xdcc-grabscher/blob/xg_3.3/README.md#api
It should cover all actions you are allowed to execute externally.

**Comment by lformella** [comment]
Fixed in [v3.3.0.0](https://github.com/lformella/xdcc-grabscher/releases/tag/v3.3.0.0)
